### PR TITLE
fix set_data permitting requires_grad=True on integer tensor

### PR DESCRIPTION
**Pull request by yuguo68**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #78436

fix https://github.com/pytorch/pytorch/issues/76008

Differential Revision: [D36800179](https://our.internmc.facebook.com/intern/diff/D36800179)

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -7433,12 +7433,12 @@ def test_dot_data_use(self):
...
                 RuntimeError,
-                # message includes both Double and Long
-                '(?=.*Double)(?=.*Long)'):
+                # message includes both Double and ComplexFloat
+                '(?=.*Double)(?=.*ComplexFloat)'):
 
             # Calls model with a LongTensor input but DoubleTensor weights
             input = torch.randn(1, 1, 1, 6, dtype=torch.double)
-            weight = torch.zeros(1, 1, 1, 3, dtype=torch.long)
+            weight = torch.zeros(1, 1, 1, 3, dtype=torch.complex64)
```
Comment (albanD) [inline test/test_torch.py:7441]: Why change this?

**Review by albanD — COMMENTED** [review]
That makes sense to me as a fix even though it is BC-breaking.
cc @soulitzer do you agree?

File: test/test_torch.py (hunk 1 above)
Comment (yuguo68) [reply test/test_torch.py:7441]: Since `torch.long` does not support grad, with this PR, this test will fail at this assignment. However, the original test fails at `out = model(input)` with a dtype mismatch error.

**Review by albanD — APPROVED** [review]
SGTM
I'll let @soulitzer take a look and merge if he's happy with this!

File: torch/csrc/autograd/python_variable.cpp [hunk 2]
```diff
@@ -768,8 +768,14 @@ int THPVariable_set_data(THPVariable *self, PyObject *data, void *unused)
...
   }
-
-  THPVariable_Unpack(self).set_data(THPVariable_Unpack(data));
+  const auto& var = THPVariable_Unpack(self);
+  const auto& other = THPVariable_Unpack(data);
+  auto self_requires_grad = var.requires_grad();
+  auto is_differentiable = isDifferentiableType(at::typeMetaToScalarType((other.dtype())));
+  THPUtils_assertRet(-1, !self_requires_grad || is_differentiable,
+      "data set to a tensor that requires gradient must be "
+      "floating point and complex dtype that support gradients");
```
Comment (soulitzer) [inline torch/csrc/autograd/python_variable.cpp:777]: nit: Maybe reword slightly to - `"data set to a tensor that requires gradients must be floating point or complex dtype"`

**Review by soulitzer — COMMENTED** [review]
Thanks for the change! Just a small nit on the wording.

File: torch/csrc/autograd/python_variable.cpp (hunk 2 above)
Comment (yuguo68) [reply torch/csrc/autograd/python_variable.cpp:777]: sure, also is it better to put this check inside set_data in variable.cpp? https://github.com/pytorch/pytorch/blob/c88367442d0e89edb101f736a06a9f965983b2c9/torch/csrc/autograd/variable.cpp#L391

Comment (yuguo68) [reply torch/csrc/autograd/python_variable.cpp:777]: let me move the check to `set_data` in variable.cpp.

**Comment by yuguo68** [comment]
@yuguo68 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36800179).

**Review by soulitzer — APPROVED** [review]
Thanks!

**Comment by soulitzer** [comment]
@yuguo68 why did you import this, are there fb-only changes that need to be made?
